rig-sqlite/src/lib.rs

```diff
@@ -0,0 +1,452 @@
+use rig::embeddings::{Embedding, EmbeddingModel};
+use rig::vector_store::{VectorStoreError, VectorStoreIndex};
+use rig::OneOrMany;
+use serde::Deserialize;
+use std::marker::PhantomData;
+use tokio_rusqlite::Connection;
+use tracing::{debug, info};
+use zerocopy::IntoBytes;
+
+#[derive(Debug)]
+pub enum SqliteError {
+    DatabaseError(Box<dyn std::error::Error + Send + Sync>),
+    SerializationError(Box<dyn std::error::Error + Send + Sync>),
+    InvalidColumnType(String),
+}
+
+pub trait ColumnValue: Send + Sync {
+    fn to_sql_string(&self) -> String;
+    fn column_type(&self) -> &'static str;
+}
+
+pub struct Column {
+    name: &'static str,
+    col_type: &'static str,
+    indexed: bool,
+}
+
+impl Column {
+    pub fn new(name: &'static str, col_type: &'static str) -> Self {
+        Self {
+            name,
+            col_type,
+            indexed: false,
+        }
+    }
+
+    pub fn indexed(mut self) -> Self {
+        self.indexed = true;
+        self
+    }
+}
+
+/// Example of a document type that can be used with SqliteVectorStore
+/// ```rust
+/// use rig::Embed;
+/// use serde::Deserialize;
+/// use rig_sqlite::{Column, ColumnValue, SqliteVectorStoreTable};
+///
+/// #[derive(Embed, Clone, Debug, Deserialize)]
+/// struct Document {
+///     id: String,
+///     #[embed]
+///     content: String,
+/// }
+///
+/// impl SqliteVectorStoreTable for Document {
+///     fn name() -> &'static str {
+///         "documents"
+///     }
+///
+///     fn schema() -> Vec<Column> {
+///         vec![
+///             Column::new("id", "TEXT PRIMARY KEY"),
+///             Column::new("content", "TEXT"),
+///         ]
+///     }
+///
+///     fn id(&self) -> String {
+///         self.id.clone()
+///     }
+///
+///     fn column_values(&self) -> Vec<(&'static str, Box<dyn ColumnValue>)> {
+///         vec![
+///             ("id", Box::new(self.id.clone())),
+///             ("content", Box::new(self.content.clone())),
+///         ]
+///     }
+/// }
+/// ```
+pub trait SqliteVectorStoreTable: Send + Sync + Clone {
+    fn name() -> &'static str;
+    fn schema() -> Vec<Column>;
+    fn id(&self) -> String;
+    fn column_values(&self) -> Vec<(&'static str, Box<dyn ColumnValue>)>;
+}
+
+#[derive(Clone)]
+pub struct SqliteVectorStore<E: EmbeddingModel + 'static, T: SqliteVectorStoreTable + 'static> {
+    conn: Connection,
+    _phantom: PhantomData<(E, T)>,
+}
+
+impl<E: EmbeddingModel + 'static, T: SqliteVectorStoreTable + 'static> SqliteVectorStore<E, T> {
+    pub async fn new(conn: Connection, embedding_model: &E) -> Result<Self, VectorStoreError> {
+        let dims = embedding_model.ndims();
+        let table_name = T::name();
+        let schema = T::schema();
+
+        // Build the table schema
+        let mut create_table = format!("CREATE TABLE IF NOT EXISTS {} (", table_name);
```

Comment:
Here, I see you are using normal string interpolation rather than proper SQL [named parameters](https://docs.rs/rusqlite/latest/rusqlite/macro.named_params.html). I saw you used parameters elsewhere in the codebase so I was wondering if there was a specific reason.

It would be best to use proper SQL practices here to avoid SQL injections (even if it would require the app developer to improperly expose fields like `table_name` to the user, I still think it's worthwhile to avoid these bugs).

Reply:
Hm afaict Sqlite doesn't support parameterized queries for table / column names, only for values

Reply:
Hmm, I see. Well I suppose we should leave a disclaimer in the documentation, perhaps in the docstring as well.